### PR TITLE
Move preallocate module to libs

### PR DESCRIPTION
**Pull request by rjernst**
The preallocate module needs access to java.io internals. However, in order to open java.io to a specific module, rather than the unnamed module as was previously done, the said module must be in the boot layer.

This commit moves the preallocate module to libs. It adds it to the main lib dir, though it does not add it as a dependency of server.

### CODE REVIEW

**Comment by rjernst** [comment]
@ChrisHegarty This still needs some cleanup, but I implemented a service for adding exports/opens to a loaded module as you suggested. See https://github.com/elastic/elasticsearch/pull/94884/commits/7dbee29f02207e12fd0e6b902a484e286f5f755d. It seems to work (mostly, there seem to be some problems with our jna interaction with modularization, but running tests with the default distro no longer have illegal access errors from preallocate). I will probably pull that service out to a separate PR next week. Not sure how to test yet, PluginsService isn't well suited to mocking these loaded services out, but i can probably at least test the export service code with unit tests and test modules.

File: libs/core/src/main/java/org/elasticsearch/jdk/ModuleQualifiedExportsService.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+/**
+ * An object that provides a callback for qualified exports and opens.
+ *
+ * Because Elasticsearch constructs plugin module layers dynamically, qualified
+ * exports are silently dropped in the boot layer. Modules that have qualified
+ * exports should implement this service so that when a qualified export
+ * module is loaded, the exporting or opening module can be informed and
+ * export or open dynamically to the newly loaded module.
+ */
+public abstract class ModuleQualifiedExportsService {
```
Comment (ChrisHegarty) [inline libs/core/src/main/java/org/elasticsearch/jdk/ModuleQualifiedExportsService.java:34]: Cool! This is exactly what I was hoping to see. 👍

File: libs/preallocate/src/main/java/org/elasticsearch/preallocate/PreallocateModuleExportsService.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.preallocate;
+
+import org.elasticsearch.jdk.ModuleQualifiedExportsService;
+
+public class PreallocateModuleExportsService extends ModuleQualifiedExportsService {
```
Comment (ChrisHegarty) [inline libs/preallocate/src/main/java/org/elasticsearch/preallocate/PreallocateModuleExportsService.java:13]: While a little unfortunate that we need to do this in all target modules, it's simple and straightforward.

Comment (ChrisHegarty) [reply libs/preallocate/src/main/java/org/elasticsearch/preallocate/PreallocateModuleExportsService.java:13]: And very much auditable. Or even generate-able if we wanted, in the future.

Comment (ChrisHegarty) [reply libs/preallocate/src/main/java/org/elasticsearch/preallocate/PreallocateModuleExportsService.java:13]: One last comment on this ( and again for the future ). What you have is good, but as an alternative we could reduce this to a single method that returns a `MethodHandles.Lookup`, that holds the PRIVATE access mode of the declaring module. The service can then use that lookup to perform invocations of caller-sensitive methods as-if it were the caller.  see https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/invoke/MethodHandles.Lookup.html#callsens

Comment (rjernst) [reply libs/preallocate/src/main/java/org/elasticsearch/preallocate/PreallocateModuleExportsService.java:13]: I like that idea, but I will leave it for the future.

**Review by ChrisHegarty — APPROVED** [review]
The modules part of the changes look good to me. 👍

**Comment by rjernst** [comment]
I attempted to write a unit test for the new exports service, but unfortunately it gets complicated since tests do not currently run modular.

**Pull request merged**